### DISCUSSION TITLE
[react-dom] `act` should only return a promise when passed a promise

### DISCUSSION THREAD

**Discussion by CreativeTechGuy** [DefinitelyTyped/DefinitelyTyped#63394] [Issues with a @types package] [answered]
Continuing the conversation from #63186.

CC: @eps1lon, @icatalina

In the linked PR, the return type of `act` was changed to always return a promise. While this is technically correct based on the implementation, this promise isn't relevant if the callback function doesn't return a promise. Returning a `Promise` when it can be safely ignored is confusing to a developer as to what they need to do and when. It is extra noisy when someone is using [@typescript-eslint/no-floating-promises](https://typescript-eslint.io/rules/no-floating-promises/) which is a recommended rule so many people probably are experiencing this combination.

I'm requesting that the types for `act` be changed to the following:

```
export function act(callback: () => VoidOrUndefinedOnly): void;
export function act<T>(callback: () => T | Promise<T>): Promise<T>;
```

This change takes the improvements from the linked PR but reverts the half of the change which caused problems.

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment, marked as answer]
Fixed https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63561. Should be released shortly.

**Comment by tkrotoff** [comment]
This is not totally fixed (can it be?)

Example:
```TypeScript
act(() => emit('routeChangeStart'));
```

[`emit()`](https://nodejs.org/docs/latest-v18.x/api/events.html#emitteremiteventname-args) returns a boolean instead of void or undefined

Thus `act` type returns `Promise<boolean>` because of

```TypeScript
export function act(callback: () => VoidOrUndefinedOnly): void;
export function act<T>(callback: () => T | Promise<T>): Promise<T>; // <===
```

Later on `@typescript-eslint` complains `act` needs to be awaited:

> Promises must be awaited [...]. eslint([@typescript-eslint/no-floating-promises](https://typescript-eslint.io/rules/no-floating-promises))

**Reply by CreativeTechGuy** [reply]
Sounds like you just need curly brackets. Let me know if this fixes your problem:

```js
act(() => {
    emit('routeChangeStart');
});
```

Since you don't care about the return value, you shouldn't be returning it to act's callback.

**Reply by tkrotoff** [reply]
Yes it does, but it makes the code less readable
